Update Filter by Stock block

tjcafferkey opened the pull request:
Update the Filter by Stock UI to be inline with the new designs.

<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->

<!-- Reference any related issues or PRs here -->

Fixes https://github.com/woocommerce/woocommerce-blocks/issues/6842

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->

#### Accessibility

<!-- If you've changed or added any interactions, check off the appropriate items below. You can delete any that don't apply. Use this space to elaborate on anything if needed. -->

- [x] I've tested using only a keyboard (no mouse)
- [x] I've tested using a screen reader
- [ ] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [ ] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

### Screenshots

<!-- If your change has a visual component, add a screenshot here. A "before" screenshot would also be helpful. -->

| Before | After |
| ------ | ----- |
|    <img width="509" alt="Screenshot 2022-08-17 at 12 33 57" src="https://user-images.githubusercontent.com/8639742/185108568-1c6875e8-10a1-4070-919d-876c42af8a89.png">    |   <img width="555" alt="Screenshot 2022-08-17 at 12 32 28" src="https://user-images.githubusercontent.com/8639742/185410884-45952fdd-14fe-4784-9ba6-c58c248630ad.png">    |

### Testing

#### Automated Tests
* [x] Changes in this PR are covered by Automated Tests.
  * [x] Unit tests
  * [x] E2E tests

#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

1. Add the Filter by Stock component to a page using the All Products block, and also a block template.
2. Check that the control panel for this block looks like the design (minus Typography)
3. On both the page and the template frontends select a checkbox and observe that it looks as it should do with the design.
4. Check that a "Reset" button appears and clicking this resets the filter again.
5. Now go to your Edit Page & Edit Template pages and enable "'Apply filters' button".
6. On the frontends again, re-select your filter options and check that they are not applied immediately until you click "Apply"
7. Check that when this block is used with the PHP template & the All Products block that regardless of whether the "Apply Filters" button is active, that clicking "Reset" immediately clears the selected filter options

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [x] Feature plugin
* [ ] Experimental

### Performance Impact

<!-- Please document any known performance impact (positive or negative) here. If negative, provide some rationale for why this is an okay tradeoff or how this will be addressed. -->

### Changelog

> Update the design of the Filter Products by Stock block.

tjcafferkey committed:
Update Filter by Stock controls and UI to include custom checkbox and settings.

tjcafferkey committed:
Merge branch 'trunk' into update/filter-by-stock-ui

tjcafferkey committed:
Merge branch 'trunk' into update/filter-by-stock-ui

tjcafferkey committed:
Include and style Filter by Stock reset button

tjcafferkey committed:
Make checkbox accessible via keyboard

tjcafferkey committed:
Merge branch 'trunk' into update/filter-by-stock-ui

tjcafferkey committed:
Update snapshot tests for Filter Products by Stock component

tjcafferkey committed:
Update E2E tests with new label value

tjcafferkey committed:
Merge branch 'update/filter-by-stock-ui' of github.com:woocommerce/woocommerce-blocks into update/filter-by-stock-ui

tjcafferkey committed:
Update label test for Apply Filters in E2E tests

tjcafferkey committed:
Specify checkbox type so styles are not overridden in the editor

tjcafferkey committed:
Update E2E test label for filter button

tjcafferkey committed:
Update label selector to search for containing text rather than exact match

tjcafferkey committed:
Escape quotation marks for label query in E2E shopper tests for Filter by Stock

tjcafferkey committed:
Correct typo on Filter by Stock controls

dinhtungdu commented on assets/js/base/components/filter-element-checkbox/style.scss:
```diff
@@ -0,0 +1,38 @@
+	display: inline-block;
+	position: relative;
+	cursor: pointer;
+	--checkbox-background-color: #{$studio-woocommerce-purple-30};
```
Should we move this to the top of this scope?

dinhtungdu replied:
Also, I only see this var is used in one place, so I'm wondering if we can use the `$studio-woocommerce-purple-30` directly?

tjcafferkey replied:
Yes you're right. In previous iterations of this PR I used it in multiple places.

dinhtungdu reviewed (commented):
@tjcafferkey The PR is testing great on my end. The main functionalities work as expected.
I found some minor issues detailed below, please take a look!

- There are some design updates that haven't been implemented in this PR:
    - The block name and description.
    - The product count now has the same color as the option label.
    - The checkbox size is 18x18px in the design. In my test for this PR, its size is actually 20x20px.
    - The focus outline for checkboxes is rounded, should we remove the radius to make it match the unfocused state style? cc @vivialice 

      <details>
      <summary>Screenshot</summary>
      <img width="411" alt="image" src="https://user-images.githubusercontent.com/5423135/185296942-38ea0685-0a04-46d2-ae43-0fcd1a7254ee.png">
      </details>
- About the Reset button, there is a case the behavior can be improved IMO:
  1. Add the block to the shop page and Enable the Apply Filter button.
  2. On the shop page, check a stock option.
  3. See the page reloads.
  4. See the Reset and Apply buttons both appear because a filter is selected now.
  5. Click on the Reset button, the page won't reload, and the Active Filters block won't update.
  6. We have to click on the Apply button to reset the filter.
=> IMO, clicking the Reset button should always reset the filter, even if the Apply filter button is enabled. cc @vivialice

Aljullu commented on assets/js/base/components/checkbox-list/index.tsx:
```diff
@@ -132,15 +133,11 @@ const CheckboxList = ( {
-								} }
-								checked={ checked.includes( option.value ) }
-								disabled={ isDisabled }
+							<FilterElementCheckbox
```
I know the styles are different (I asked about that here: pdnLyh-1SH-p2#comment-1245), but I wonder if we could reuse the [`CheckboxControlProps` component](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/packages/checkout/components/checkbox-control/index.tsx) anyway. It might require adding a prop to control the different styles, but that way we would need to maintain one less component. Do you think that would work, @tjcafferkey?

tjcafferkey replied:
Thanks @Aljullu I'll give this a look into as well!

tjcafferkey committed:
Replace custom checkbox with CheckboxControl component from the checkout package.

tjcafferkey committed:
Remove opacity on product count for checkbox label so its the same color as the label.

tjcafferkey committed:
Update snapshot tests with new component usage for filter by stock

tjcafferkey committed:
Update block title from Filter Products by Stock to Filter by Stock

tjcafferkey commented:
@dinhtungdu thanks for your review:

>The block name and description.

I have updated the block name, but the description seems as it should be?

>The product count now has the same color as the option label.

This has been sorted now.

>The checkbox size is 18x18px in the design. In my test for this PR, its size is actually 20x20px.
>The focus outline for checkboxes is rounded, should we remove the radius to make it match the unfocused state style? cc @\vivialice

We have decided to use the `CheckboxControl` from `checkout` for consistency across the UI. I will create a separate issue to move this component to a place that makes more sense seeing as its now used _outside_ of checkout too. Haven't done this as part of this PR to limit scope and regression testing.

>About the Reset button, there is a case the behavior can be improved IMO

Hmm, I am not sure but I totally understand your point. I will leave this up to @vivialice to decide upon.

@Aljullu I have addressed your feedback also, and the PR is updated with the new checkbox component. Good catch! I should have known one should have existed.

tjcafferkey committed:
Merge branch 'trunk' into update/filter-by-stock-ui

vivialice commented:
> The focus outline for checkboxes is rounded, should we remove the radius to make it match the unfocused state style? cc @vivialice

Yes! 👍 

> clicking the Reset button should always reset the filter, even if the Apply filter button is enabled. 

Correct. I also had forgotten the whole page refreshes. I remember during one of the past iterations we spoke about loading states - has this been updated or accounted for if not?

tjcafferkey commented:
>I remember during one of the past iterations we spoke about loading states - has this been updated or accounted for if not?

@vivialice If you mean the loading placeholders, this will be done in a separate pull request.

>Correct. 

Thanks! I will update this behaviour now then

tjcafferkey committed:
Update reset button behaviour to clear selections immediately regardless of whether the Apply Filters button is active or not.

tjcafferkey committed:
Merge branch 'trunk' into update/filter-by-stock-ui

tjcafferkey committed:
Pass option value as ID to satisfy E2E tests

tjcafferkey committed:
Update snapshot tests for stock filter

dinhtungdu reviewed (commented):
@tjcafferkey The update looks good, but I found a minor issue with the reset button. Other than that, this is LGTM!

Steps to reproduce:

1. Enable the Apply filter button.
2. On the front end, select a stock filter.
3. Click on the reset button.
4. See the filter doesn't reset.

tjcafferkey committed:
Reset checked state on reset for when they have filters checked but not applied.

tjcafferkey commented:
Good catch @dinhtungdu Ive addressed that here https://github.com/woocommerce/woocommerce-blocks/pull/6883/commits/71ad0dffdf4c3cfa945175693d2e6ddf87ca86e3

dinhtungdu reviewed (approved):
Thanks for the lighting fast update! 🚀
